### ISSUE TITLE
Improve dependency updates

### ISSUE THREAD

**Issue by planetf1** [odpi/egeria#4014]
We have many dependencies as part of egeria, and over time the maven definitions have also included more and more transitive dependency versions to address specific issues, or currency/security .

We are using dependabot to manage dependencies, which will result in a PR being proposed for each available update 

We have recently added a parallel gradle build, with a view to potentially moving to it in future due to hugely increased performance, and this has followed the same pattern, and expressed the same dependencies.

Over time more varied code has been added to the egeria project

The result is a fairly heavy stream of dependabot update proposals which can be cumbersome to manage.

Docs on the config options are at https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/configuration-options-for-dependency-updates#ignore . We can for example change the number of PRs open at any time, and change pattern matching for include/exclude, or the time the scan takes place

We should consider how this might be handled better

* Dependabot can no longer seemingly be switched to security updates only, though packages can be excluded by name
* Tools like 'mvn versions:display-dependency-updates' may offer an alternative, but have their own set of issues
* auto-merging of approved dependabot updates was available on prior dependabot versions but is no longer supported directly. Other github actions may be abe to do this - meaning we could just 'approve' and let the merges happen in the background
* Manual application of updates is possible (review the list, code the change, push a PR) but require more time - but will mean a much quicker integration of updates without many PRs
* Reducing the diversity within a single repo should focus the updates - today our egeria repo has java code, javascript, docker images, plugins for 3rd party technology - there is intent to move some of these out which should make the updates to core egeria clearer, but of course will add the dependabot workload to other areas. This reduction will also likely reduce build time making the impact of a PR merge easier
* We could switch off the gradle dependabot updates until gradle is fully ready - but in doing so the gradle build will likely degrade and/or fail over time 
* We could try to accelerate a move to gradle -- the build is much much faster, especially for incremental checks, again reducing the impact of the PRs (and ultimately allowing maven dependency updates to be stopped)
* We could allow dependency updates to skip full PR validation. However bugs have frequently been found in unit/fvts including caused by dependency updates, so this seems undesirable. perhaps small tweaks like not running a maven build if only gradle build files have been changed (but this could still cause a failure - for example if the new build change breaks RAT checks)
* Encouraging broader term participation in reviewing and actioning version PRs

It's not clear there's any obvious steps that will have an impressive immediate effect without comprimising other areas.

In summary my thoughts are:
 * We need to keep the code up to date, and need to keep on top of security, so leave dependabot active
 * Trying to get some kind of auto-merge setup, ideally so that approved dependabot PRs are merged overnight or at the weekend automatically (and increase open PRs, perhaps switch schedule to weekly) 
 * Try and move faster with gradle work due to excellent potential benefits
 * Open up discussion on more repo splitting. Samples already discussed, but UI/javascript is another candidate.

Opening for discussino

**Comment by grahamwallis** [comment]
@planetf1 - Good discussion.

My initial reaction is that we should retain dependant and we should not drop our guard on PR validation. Perhaps though we should reduce the cadence of dependant checks to once per release, at the beginning of that release cycle?

**Comment by planetf1** [comment]
I agree on reducing the frequency of checks. I was initially thinking weekly, but perhaps monthly is indeed better as it gives us the best opportunity to make significant changes early in the cycle, with sufficient time for finding any regressions as we progress towards a release at the end of the month.

The other thought (which I'm trying now) is to, where possible, create a branch of master, cherry-pick all of dependabot's commits, and then create a single PR - or at least a small number to ease the PR process.  At some point it might be possible to code an action to assist in this, though it's a relatively quick task.

Dependabot's documentation states that if we go 'monthly' it HAS to be the first day of the month: https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/configuration-options-for-dependency-updates

I think this works well with our schedule in any case

I believe security fixes will still be alerted via github 'security' tab outside this cycle - which we want, and of course we can manually make changes as needed.

**Comment by planetf1** [comment]
In this case I would also open-up the 'limit' on number of outstanding PRs to a large number, so that we gather all the updates at once. 

@grahamwallis @davidradl How do you feel about the checks on presentation server? Currently we have a backlog of updates needed. Currently they are throttled to 1 intentionally to allow the team to get up to speed, but I think it would be better to open up and be consistent with the rest of our checks for this next release . 

However if we do someone from the UI team needs to action promptly & triage the reasonableness of proposals - I know npm dependencies can be especially troublesome, plus it will be the first time we do a significant update.

A comprimise may be to defer this for the UI for one further month.

**Comment by grahamwallis** [comment]
@planetf1 - Maybe instead of going from 1 to infinity, we could grow the limit to 5 or 10 for a short time until we have built up momentum?

**Comment by planetf1** [comment]
> @planetf1 - Maybe instead of going from 1 to infinity, we could grow the limit to 5 or 10 for a short time until we have built up momentum?

Specifically for the UI? Yes makes sense if you're ok to take that on this month ... You make a call on the number and then we can try and increase for next month. For the base I think a large number is ok (eek).

**Comment by planetf1** [comment]
If we can agree on this I'd be inclined to do it soon/before Sun then all is in place without rework?
@cmgrote @mandy-chessell  ok?

In summary
 - Dependabot becomes monthly, first of month
 - high limit on #s (ie 100) for most areas (Docker, java)
 - Low limit on javascript (5 - tbc @grahamwallis ) to test the process
 - intent to cherry-pick commits at beginning of month into a clean PR - or at least a small number, to reduce overhead on build (cc: @cmgrote )

**Comment by planetf1** [comment]
To clarify, the points about
 - repo splitting
 - gradle build
 - automerge

All have their own benefits, but are covered in other PRs (on their own merits) so will add notes there (later)

**Comment by grahamwallis** [comment]
@planetf1 - Yes I'm happy to look at the UI ones - I'll handle all those I can but may need to triage and hand-off some of them depending on what is impacted.

I'd prefer to keep the number small initially so let's go with 5 and scale up as we can.

**Comment by planetf1** [comment]
repo -> #458 (ranger) , #2739 (samples)- and note we already have #odpi/egeria-ui and #odpi/egeria-k8s-operator
gradle -> #3370 
automerge -> #2453

**Comment by planetf1** [comment]
Suggest we close this now given the improvements or noted actions above.
